src/layers/domain/core/questionnaire.py

```diff
@@ -1,75 +1,149 @@
-            raise DuplicateError(f"Question exists: {name}")
-        result = Question(name=name, type=type, multiple=multiple)
-        self._questions.append(result)
-        return result
+        if name in self.questions:
+            raise DuplicateError(f"Question '{name}' already exists.")
 
-    def remove_question(self, name: str):
-        """
-        Removes a question from the questionnaire.  Once a questionnaire is
-        locked this method will throw an error.
-        """
-        raise NotImplementedError()
+        # Validate choices are of the same type as the question type
+        if choices is not None and not all(
+            isinstance(choice, type) for choice in choices
+        ):
+            raise ValueError(
+                f"Choices must be of the same type as the question type: {type}"
+            )
 
-    def revise(self) -> "Questionnaire":
-        """
-        Creates an exact duplicate of the questionnaire, which is the only way
-        to edit questionnaires once they become locked.
-        """
-        raise NotImplementedError()
+        question = Question[type](
+            name=name,
+            type=type,
+            multiple=multiple,
+            validation_rules=validation_rules,
+            choices=choices,
+        )
+
+        self.questions[name] = question
+        return question
+
+
+class QuestionnaireResponse(BaseModel):
+    """
+    Validates questionnaire responses against questionnaire questions
+    """
+
+    questionnaire: Questionnaire
+    responses: list[tuple[str, list]]
+
+    @validator("responses", each_item=True)
+    def validate_responses(
+        response: tuple[str, list], values: dict[str, Questionnaire]
+    ):
+        questionnaire = values.get("questionnaire")
+        question_name = response[0]
+        if questionnaire is not None:
+            questionnaire_name = questionnaire.name
+            question = questionnaire.questions.get(question_name)
+            if question is None:
+                raise InvalidResponseError(
+                    f"Unexpected answer for the question '{question_name}'. The questionnaire '{questionnaire_name}' does not contain this question."
+                )
+            validate_response_against_question(response=response, question=question)
+        return response
+
+
+def validate_response_against_question(response: tuple[str, list], question: Question):
+    answers = response[1]
+
+    if not question.multiple and len(answers) > 1:
+        raise InvalidResponseError(
+            f"Question '{question.name}' does not allow multiple responses. Response given: {answers}."
+        )
+
+    errors = (
+        []
+    )  # accumulate errors here for multianswer question and raise under a single ValueError
+    for answer in answers:
+        if not isinstance(answer, question.type):
+            errors.append(
+                f"Question '{question.name}' expects type {question.type}. Response given: {answer}."
```

Comment:
just to help debugging:

```suggestion
                f"Question '{question.name}' expects type {question.type}. Response '{answer}' is of type '{type(answer)}'."
```